design/baremetal-operator/bmh-v1beta1.md

```diff
@@ -0,0 +1,148 @@
+incompatible changes are needed based on the recent development both in Metal3
+itself and in Ironic.
+
+For example, Ironic is planning a complete removal of MD5 support for
+checksums.
+
+### Goals
+
+- Clean up issues in the BareMetalHost API that have accumulated during
+  the usage of `v1alpha1`.
+
+- Make the API version reflect the actual support status of the API (see
+  [alternatives](#Alternatives) for details).
+
+### Non-Goals
+
+- Stabilize the BareMetalHost API completely (as `v1`).
+
+## Proposal
+
+Copy the current API types definition into a new package `v1beta1` with some
+modifications.
+
+## Design Details
+
+The following addition should be done *before* the version is created:
+
+- Add a `Spec.Architecture` field to override architecture in the profile.
+  If not provided, BMO will use its own architecture as the default.
+
+The following changes will be done to the new version:
+
+- Remove MD5 support from `Image.ChecksumType`. When no `ChecksumType` is
+  provided, do not populate the `os_image_hash_algo` field and let Ironic
+  guess the algorithm.
+
+  The same change is applied to CAPM3.
+
+- In `RootDeviceHints`:
+
+  Rename `DeviceName` to `DevicePath` to better reflect what it actually is
+  (especially after introducing support for `/dev/disk/by-path`).
+
+  Rename `Model` and `Vendor` to `ModelContains` and `VendorContains` to
+  reflect the fact that these fields use the `<in>` operator.
+
+- Remove `HardwareProfile` and don't do any matching.
+
+- Remove `HardwareDetails` from the `Status` in favour of the already
+  introduced `HardwareData` CR.
+
+- Rename `online` to `poweredOn`.
+
+  The word *online* is ambiguous here: it may be understood to refer to
+  provisioning status. Furthermore, the corresponding field in the `Status`
+  is already called `poweredOn`.
+
+### Implementation Details/Notes/Constraints
+
+Both `v1alpha1` and the proposed `v1beta1` resources will be possible
+to handle with the same controller code since the differences are mostly
+cosmetic.
+
+### Risks and Mitigations
+
+None
+
+### Work Items
+
+- Refactor the BMO code to avoid referring directly to the current version.
+- Add `Architecture` to the current version.
+- Copy type definitions and modify them.
+- Create a conversion webhook to change between versions.
+- Make a new BMO release and communicate the change to the community.
+- In the future: create a BMO branch and delete `v1alpha1` afterwards.
+
+### Dependencies
+
+None
+
+### Test Plan
+
+We'll need to test both API versions. Since we already have tests for different
+versions of CAPM3, we will tie them to different versions of BMH. For instance,
+the current version of CAPM3 will switch to the new API version, while all
+stable branches will keep using `v1alpha1`. We'll only need to add one of
+such stable jobs to BMO.
+
+### Upgrade / Downgrade Strategy
+
+The conversion webhook will take care of in-flight CR upgrade:
```

Comment:
Also reading HardwareDetails from the separate CRD when converting back to v1alpha1.

Reply:
I don't believe it's possible, but it's probably also not required if 1) the storage version is v1alpha1, 2) the controller keeps syncing the separate CR and Status.